src/main/scala/com/fulcrumgenomics/bam/OverlappingBasesConsensusCaller.scala

```diff
@@ -0,0 +1,250 @@
+  *
+  * This will iterate through the mapped bases that overlap between the read and mate in a read pair.  If the read and
+  * mate agree at a given reference position, then read and mate base will not change and the base quality returned
+  * is controlled by `maxQualOnAgreement`. If they disagree at a given reference position, then the base and quality
+  * returned is controlled by `onlyMaskDisagreements`.
+  *
+  * @param onlyMaskDisagreements if the read and mate bases disagree at a given reference position, true to mask (make
+  *                              'N') the read and mate bases, otherwise pick the base with the highest base quality and
+  *                              return a base quality that's the difference between the higher and lower base quality.
+  * @param maxQualOnAgreement    if the read and mate bases agree at a given reference position, true to for the
+  *                              resulting base quality to be the maximum base quality, otherwise the sum of the base
+  *                              qualities.
+  */
+class OverlappingBasesConsensusCaller(onlyMaskDisagreements: Boolean = false,
+                                      maxQualOnAgreement: Boolean = false) {
+  private val NoCall: Byte = 'N'.toByte
+  private val NoCallQual: PhredScore = PhredScore.MinValue
+
+  private val r1BasesBuilder = Array.newBuilder[Byte]
+  private val r1QualsBuilder = Array.newBuilder[PhredScore]
+  private val r2BasesBuilder = Array.newBuilder[Byte]
+  private val r2QualsBuilder = Array.newBuilder[PhredScore]
+  private val r1Builders     = Seq(r1BasesBuilder, r1QualsBuilder)
+  private val r2Builders     = Seq(r2BasesBuilder, r2QualsBuilder)
+
+  /** Consensus calls the overlapping bases if and only if the template is a paired end where both ends map with at
+    * least one base overlapping.
+    *
+    * @param template the template to potentially correct.
+    * @return summary statistics about how many bases were examined and modified
+    */
+  def call(template: Template): CorrectionStats = (template.r1, template.r2) match {
+    case (Some(r1), Some(r2)) if r1.mapped && r2.mapped && r1.matesOverlap.contains(true) => call(r1, r2)
+    case _ => CorrectionStats(0, 0, 0, 0)
+  }
+
+  /** Consensus calls the overlapping bases if and only if both ends map with at least one base overlapping.
+    *
+    * @param r1 the first read in the pair
+    * @param r2 the second read in the pair
+    * @return summary statistics about how many bases were examined and modified
+    */
+  def call(r1: SamRecord, r2: SamRecord): CorrectionStats = {
+    require(r1.mapped && r2.mapped && r1.paired && r2.paired && r1.name == r2.name && r1.matesOverlap.contains(true))
+
+    // Clear and resize the builders
+    r1Builders.foreach { builder =>
+      builder.clear()
+      builder.sizeHint(r1.length)
+    }
+    r2Builders.foreach { builder =>
+      builder.clear()
+      builder.sizeHint(r2.length)
+    }
```

Comment:
I'm _so_ confused.  Why are you using builders here?  Looking at the code below it becomes really complicated because you're copying slices of the bases/quals into the builder based on what you've skipped over etc.  Why not do the edits in place in the `r1.bases` and `r1.quals` etc.?  That way you can ignore/skip over bases that are not overlapped and/or not aligned to the reference, and just tweak the bases that are aligned to the reference?  

It feels like this would be both simpler, _and_ lead to way less temporary object creation?

Reply:
done